File: utils/parse_json.py
"""
utils/parse_json.py — Safe JSON parser for LLM responses.

Handles markdown fences, trailing commas, and noisy responses.
"""

import json
import re


def parse_json(raw: str) -> dict:
    """
    Parse JSON from an LLM response, handling common formatting issues.

    Args:
        raw: Raw string from LLM response

    Returns:
        Parsed dict

    Raises:
        ValueError: If parsing fails after cleanup
    """
    if not isinstance(raw, str) or not raw.strip():
        raise ValueError("parse_json: received empty or non-string input")

    cleaned = raw.strip()

    # Strip markdown code fences
    cleaned = re.sub(r"^```(?:json)?\s*\n?", "", cleaned, flags=re.IGNORECASE)
    cleaned = re.sub(r"\n?```\s*$", "", cleaned)
    cleaned = cleaned.strip()

    # Remove trailing commas before closing brackets/braces
    cleaned = re.sub(r",\s*([}\]])", r"\1", cleaned)

    try:
        return json.loads(cleaned)
    except json.JSONDecodeError as first_error:
        # Try extracting first JSON object
        match = re.search(r"\{[\s\S]*\}", cleaned)
        if match:
            try:
                extracted = re.sub(r",\s*([}\]])", r"\1", match.group(0))
                return json.loads(extracted)
            except json.JSONDecodeError:
                pass

        raise ValueError(
            f"parse_json: failed to parse LLM response. "
            f"First 200 chars: \"{raw[:200]}\". "
            f"Error: {first_error}"
        )

File: utils/parseJSON.js
/**
 * Safe JSON parser with controlled error handling.
 * Strips markdown fences and trailing commas before parsing.
 *
 * @param {string} raw - Raw string from LLM response
 * @returns {object} Parsed JSON object
 * @throws {Error} If parsing fails after cleanup
 */
function parseJSON(raw) {
  if (typeof raw !== "string" || raw.trim().length === 0) {
    throw new Error("parseJSON: received empty or non-string input");
  }

  let cleaned = raw.trim();

  // Strip markdown code fences (```json ... ``` or ``` ... ```)
  cleaned = cleaned.replace(/^```(?:json)?\s*\n?/i, "").replace(/\n?```\s*$/, "");
  cleaned = cleaned.trim();

  // Remove trailing commas before closing brackets/braces
  cleaned = cleaned.replace(/,\s*([}\]])/g, "$1");

  try {
    return JSON.parse(cleaned);
  } catch (firstError) {
    // Try extracting first JSON object from the string
    const objectMatch = cleaned.match(/\{[\s\S]*\}/);
    if (objectMatch) {
      try {
        const extracted = objectMatch[0].replace(/,\s*([}\]])/g, "$1");
        return JSON.parse(extracted);
      } catch (_) {
        // fall through
      }
    }

    throw new Error(
      `parseJSON: failed to parse LLM response. ` +
        `First 200 chars: "${raw.slice(0, 200)}". ` +
        `Error: ${firstError.message}`
    );
  }
}

module.exports = { parseJSON };
